### PR TITLE
Adding new parameter to `generate`:  `max_time`.

### PR DESCRIPTION
**Pull request by Narsil**
Generation by tokens number is sometimes a bit clunky because we don't
know how many tokens are good enough or even how many tokens are in
the payload (for pipelines users for instance). This leads to hard
to understand behavior.

This PR proposes a new argument `max_time` which is a float of seconds
for the allowed time for `generate` to run on.
Ideally combinations of `max_tokens=None`, `max_time=2` could be used to
generate as many tokens as possible within time budget.

NB: Another possible approach consists of passing a callback to `generate`
  putting the caller in charge of the actual decision of when to stop
  generating tokens. It opens the door to 'which args should we pass'
  to this callback. It's hard to imagine other use-cases for this
  early stopping behavior than time (that are not already covered by
  parameters of generate)

# What does this PR do?

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@patrickvonplaten @LysandreJik
@jplu 

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors which may be interested in your PR.

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
Continuing a bit the discussion we had offline for others to chime in.

After quite some discussion and thinking, I see the following problem:

- I don't want to clutter `generate()` with if-statements anymore as it's done a bit in this PR, but rather make use of tools like `LogitsProcessor`. Now @Narsil you gave me some very good arguments to why just adding a `LogitsProcessor` that forces to generate EOS is not good enough (Some models don't have the EOS token & we don't always want to have EOS at the end of the sentence). So I would propose the following solution that we should then also use to deprecate `max_length` from the "lower" generate methods like `greedy_search`, `sample`, ...

Analogoues to `LogitsProcessor` and `LogitsProcessorList` we create a new logit called `StoppingCriteria` and `StoppingCriteriaList` which would look as follows: 

```python
class StoppingCriteriaList(list):
    def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor, **kwargs) -> torch.FloatTensor:
        stopping = False
        for processor in self:
           stopping = stopping or processor(input_ids, scores)
        return stopping
```

and a `MaxTimeStopping` class as follows:

```python
class MaxTimeStopping(StoppingCriteria):
    def __init__(self, max_time):
        self.start_time = time.time()
        self.max_time = max_time

    def __call__(self, *args):
        if time.time() - self.start_time > self.max_time:
            return True
       return False
```

The same way we can create a `MaxLengthStopping` class as follows:

```python
class MaxLengthStopping(StoppingCriteria):
    def __init__(self, max_length):
        self.start_time = time.time()
        self.max_length = max_length

    def __call__(self, input_ids, *args):
       if input_ids.shape[-1] > self.max_length:
           return True
       return False
```

Then we can add create a `stopping_criteria` list object in generate along side creating the `logits_processor` list object and pass it to the submodules. In each submodule we would then do something like 

```
if stopping_criteria(input_ids, scores):
    break;
```

I would then also deprecate the `max_length` as an input parameter to `greedy_search` etc and add a `stopping_criteria` list object instead. 

This new approach would open the way for more fancy stopping criteria. E.g. at the moment `max_length` defines the number of total tokens (passed tokens + generated tokens) instead of just the generated tokens which is very hard to change in terms of backwards compatibility. Lots of people have complained about that. With this approach, one could easily make a new `MaxGeneratedTokenStopping` class that would then take over.

This will require a rather big change, so I'd be very glad if @LysandreJik and @sgugger you can give your opinion here before proceeding.

**Comment by LysandreJik** [comment]
Thanks for the thoughtful explanation, this makes a lot of sense. I'm very down to continue the modular approach we have with processors, the new `StoppingCriteria` you propose seems like the way to go. It's good that it keeps the extensibility of the generation methods while not complexifying the generate method itself.

**Comment by sgugger** [comment]
Agreed with both of you, this `StoppingCriteria` class seems like a good idea!

**Comment by Narsil** [comment]
@patrickvonplaten @LysandreJik 
Do you mind a second review ?

I think this PR is actually ready.

The TF code (which was my main concern) doesn't seem to use LogitsProcessor nor to be tested, so I figured leaving `max_time` is ok. I could also simply remove it to make sure I don't break things.

File: src/transformers/generation_stopping_criteria.py
Comment (patrickvonplaten) [inline src/transformers/generation_stopping_criteria.py:11]: Could we get some docstring here?

File: src/transformers/generation_utils.py
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:650]: Can we let `max_time` default to `None` in `configuration_utils.py`?

File: src/transformers/generation_utils.py
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:621]: Can we add some:

```
max_length = max_length if max_length is not None else max_length
```

same for `max_time` 

and remove where it's done earlier?

File: src/transformers/generation_utils.py
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:1138]: Can we remove `max_length` from the doctsrings and deprecate it?

File: src/transformers/generation_utils.py
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:1165]: Can we add a deprecation warning here for max_length?

File: src/transformers/generation_utils.py
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:1283]: Cool!

File: src/transformers/generation_stopping_criteria.py
Comment (patrickvonplaten) [inline src/transformers/generation_stopping_criteria.py:7]: I think I'd prefer to make this an abstract `ABC` class, just like it's done for the `LogitsProcessor` class

File: tests/test_generation_stopping_criteria.py
Comment (patrickvonplaten) [inline tests/test_generation_stopping_criteria.py:17]: great tests!

**Review by patrickvonplaten — COMMENTED** [review]
Great! Thanks a lot for tackling this PR!

I'm quite happy with the design :-) 

Can we:

1) Add some docstring for the classes and add those classes to the docs? Give it a new section in `docs/source/internal/generation_utils.rst`

2) Deprecate the `max_length` function input argument for all `greedy_search`, `beam_search` and update the docstring and tests using the new `StoppingCriteriaList` instead

3) Change the `class StoppingCriteria` to an abstract class so keep the design as close as possible to the one in `LogitsProcessor...`

4) Delete the functionality for TF. If it would be ok for you, I'd like to just add this functionality for PyTorch for now since TF needs a big refactor before adding more features IMO

File: src/transformers/generation_utils.py
Comment (Narsil) [reply src/transformers/generation_utils.py:1138]: I can't because of this:

https://github.com/huggingface/transformers/blob/master/src/transformers/generation_utils.py#L1418

I'm not sure how to solve this. This probably leads to a significant perf loss if we remove it (not reallocating is important IMO, but that as usual would need some actual testing to make sure.).

In any case, I felt that this PR was already large enough as is, and that deprecating `max_length` and handling its usage (here and in `beam_sample` and `beam_search`) is something a bit more involved and would require some discussions which should not prevent (IMHO) to pass this PR

File: src/transformers/generation_stopping_criteria.py
Comment (Narsil) [reply src/transformers/generation_stopping_criteria.py:7]: I don't really understand why you add `ABC`, because you don't use `@abstractmethod` (or others).

What difference does it make ?

File: src/transformers/generation_stopping_criteria.py
Comment (Narsil) [reply src/transformers/generation_stopping_criteria.py:11]: Sure !

File: src/transformers/generation_utils.py
Comment (Narsil) [reply src/transformers/generation_utils.py:650]: Do you really think we should add this, to the configuration files ?
Generation time seems like something caller dependent, not really model dependent. (I mean why would a model limit it's generation time by default to X seconds? The results would depend on GPU/CPU exact hardware etc...)

Comment (Narsil) [reply src/transformers/generation_utils.py:650]: I can really add it, but it feels odd

File: src/transformers/generation_utils.py
Comment (Narsil) [reply src/transformers/generation_utils.py:621]: I can't for `max_length` (because of the other `max_length` usages) and deprecation. 
I'll do if you still want `max_time` to be in the config.

**Comment by Narsil** [comment]
> Great! Thanks a lot for tackling this PR!
> 
> I'm quite happy with the design :-)
> 
> Can we:
> 
>     1. Add some docstring for the classes and add those classes to the docs? Give it a new section in `docs/source/internal/generation_utils.rst`

Done, I also added the import statements within `src/transformers/__init__.py`. Is there any other place I should think of ?

> 
>     2. Deprecate the `max_length` function input argument for all `greedy_search`, `beam_search` and update the docstring and tests using the new `StoppingCriteriaList` instead

This is something harder to do because of some other usages of `max_length`. (see other comment). I think it should belong in another PR, because this one is already a bit large. And it would require other kinds of care (regarding performance at least).

What do you think ?

> 
>     3. Change the `class StoppingCriteria` to an abstract class so keep the design as close as possible to the one in `LogitsProcessor...`

Done. shouldn't they actually contain `@abstractmethod` ?

> 
>     4. Delete the functionality for TF. If it would be ok for you, I'd like to just add this functionality for PyTorch for now since TF needs a big refactor before adding more features IMO

Ok.

File: src/transformers/generation_utils.py
Comment (patrickvonplaten) [reply src/transformers/generation_utils.py:666]: Yeah, I agree with you here! Let's not add it to the config for now - the general philosophy for generate used to be to just add everything to the config, but as you pointed out, it really doesn't make much sense for config params like `max_time`

File: src/transformers/generation_utils.py
Comment (patrickvonplaten) [reply src/transformers/generation_utils.py:1154]: One thing we have to do in the PR before merging IMO is to make sure we ensure full backward compatibility though. If I understand correctly, if one directly uses `greedy_search` at the moment instead of calling `generate()`, then `stopping_criteria` has to be provided or else the generation would not stop. 

Could we for now check if a `MaxLengthCriteria` is in the list atnd if yes and if it's max_length is != the given provided max_length => overwrite it? And if no `MaxLengthCriteria` is provided, we'll just add it?

File: src/transformers/generation_utils.py
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:1303]: IMO, we should add the following lines in **all** sub-generate methods for backwards compatibility:

For backwards compatibility IMO, we should add something like:

```python
if max_length is not None and # check that no max length logits processor is in StoppingCriteriaList:
   stopping_criteria.append(MaxLengthCriteria(max_length))
```

Comment (patrickvonplaten) [reply src/transformers/generation_utils.py:1303]: Regarding the tests, I think we should check that passing `max_length` and no `StoppingCriteriaList` list object yields the same result as not passing `max_length`, but passing a `StoppingCriteriaList` with a `MaxLengthCriteria`

Comment (Narsil) [reply src/transformers/generation_utils.py:1309]: But it's impossible, we currently still require `max_length`. So no one can call without passing `max_length`. It won't work (yet).

What I've done is adding a check that if some `stopping_criteria` along a `max_length` and `stopping_critieria.max_length` does not exist or is different from `max_length` I would :
- Warn if different
- Silently add MaxLengthCriteria if missing
- No op if both exists and are the same.

File: src/transformers/generation_utils.py
Comment (Narsil) [reply src/transformers/generation_utils.py:1159]: It does.

File: src/transformers/generation_utils.py
Comment (patrickvonplaten) [reply src/transformers/generation_utils.py:1309]: Ok, yeah I'm fine with it!

**Review by patrickvonplaten — APPROVED** [review]
Thanks a lot for the PR! Very clean design :-)

Can we run the T5 and Bart slow tests once, just to be sure everything works as expected:

```bash
RUN_SLOW=1 pytest tests/test_modeling_t5.py
RUN_SLOW=1 pytest tests/test_modeling_bart.py
```

If they all pass, then it's good to be merged for me

**Comment by Narsil** [comment]
T5 Also passed, but had a OOM crash on my local machine.
```
================================================================================================================================== test session starts ===================================================================================================================================
platform linux -- Python 3.8.5, pytest-6.1.1, py-1.9.0, pluggy-0.13.1
rootdir: /home/nicolas/src/transformers
plugins: forked-1.3.0, xdist-2.1.0
collected 114 items                                                                                                                                                                                                                                                                      

tests/test_modeling_bart.py ........................................ssss.......................^[[A..............................ssss............. [100%]

============================================================= warnings summary =============================================================
.venv/lib/python3.8/site-packages/tensorflow/python/autograph/utils/testing.py:21
  /home/nicolas/src/transformers/.venv/lib/python3.8/site-packages/tensorflow/python/autograph/utils/testing.py:21: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp

tests/test_modeling_bart.py::BartModelTest::test_torchscript
tests/test_modeling_bart.py::BartModelTest::test_torchscript_output_attentions
tests/test_modeling_bart.py::BartModelTest::test_torchscript_output_hidden_state
  /home/nicolas/src/transformers/.venv/lib/python3.8/site-packages/torch/nn/functional.py:1897: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
    assert padding_idx < weight.size(0), "Padding_idx must be within num_embeddings"

tests/test_modeling_bart.py::BartModelTest::test_torchscript
tests/test_modeling_bart.py::BartModelTest::test_torchscript_output_attentions
tests/test_modeling_bart.py::BartModelTest::test_torchscript_output_hidden_state
tests/test_modeling_bart.py::BartStandaloneDecoderModelTest::test_torchscript
tests/test_modeling_bart.py::BartStandaloneDecoderModelTest::test_torchscript_output_attentions
tests/test_modeling_bart.py::BartStandaloneDecoderModelTest::test_torchscript_output_hidden_state
  /home/nicolas/src/transformers/src/transformers/models/bart/modeling_bart.py:213: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
    assert attn_weights.size() == (

tests/test_modeling_bart.py::BartModelTest::test_torchscript
tests/test_modeling_bart.py::BartModelTest::test_torchscript_output_attentions
tests/test_modeling_bart.py::BartModelTest::test_torchscript_output_hidden_state
tests/test_modeling_bart.py::BartStandaloneDecoderModelTest::test_torchscript
tests/test_modeling_bart.py::BartStandaloneDecoderModelTest::test_torchscript_output_attentions
tests/test_modeling_bart.py::BartStandaloneDecoderModelTest::test_torchscript_output_hidden_state
  /home/nicolas/src/transformers/src/transformers/models/bart/modeling_bart.py:220: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
    assert attention_mask.size() == (

tests/test_modeling_bart.py::BartModelTest::test_torchscript
tests/test_modeling_bart.py::BartModelTest::test_torchscript_output_attentions
tests/test_modeling_bart.py::BartModelTest::test_torchscript_output_hidden_state
tests/test_modeling_bart.py::BartStandaloneDecoderModelTest::test_torchscript
tests/test_modeling_bart.py::BartStandaloneDecoderModelTest::test_torchscript_output_attentions
tests/test_modeling_bart.py::BartStandaloneDecoderModelTest::test_torchscript_output_hidden_state
  /home/nicolas/src/transformers/src/transformers/models/bart/modeling_bart.py:252: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
    assert attn_output.size() == (

tests/test_modeling_bart.py::BartModelTest::test_torchscript
tests/test_modeling_bart.py::BartModelTest::test_torchscript_output_attentions
tests/test_modeling_bart.py::BartModelTest::test_torchscript_output_hidden_state
tests/test_modeling_bart.py::BartStandaloneDecoderModelTest::test_torchscript
tests/test_modeling_bart.py::BartStandaloneDecoderModelTest::test_torchscript_output_attentions
tests/test_modeling_bart.py::BartStandaloneDecoderModelTest::test_torchscript_output_hidden_state
  /home/nicolas/src/transformers/src/transformers/models/bart/modeling_bart.py:856: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
    if input_shape[-1] > 1:

-- Docs: https://docs.pytest.org/en/stable/warnings.html
========================================= 106 passed, 8 skipped, 28 warnings in 562.15s (0:09:22) ==========================================
```

**Comment by patrickvonplaten** [comment]
> T5 Also passed, but had a OOM crash on my local machine.
> 
> ```
> ================================================================================================================================== test session starts ===================================================================================================================================
> platform linux -- Python 3.8.5, pytest-6.1.1, py-1.9.0, pluggy-0.13.1
> rootdir: /home/nicolas/src/transformers
> plugins: forked-1.3.0, xdist-2.1.0
> collected 114 items                                                                                                                                                                                                                                                                      
> 
> tests/test_modeling_bart.py ........................................ssss.......................^[[A..............................ssss............. [100%]
> 
> ============================================================= warnings summary =============================================================
> .venv/lib/python3.8/site-packages/tensorflow/python/autograph/utils/testing.py:21
>   /home/nicolas/src/transformers/.venv/lib/python3.8/site-packages/tensorflow/python/autograph/utils/testing.py:21: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
>     import imp
> 
> tests/test_modeling_bart.py::BartModelTest::test_torchscript
> tests/test_modeling_bart.py::BartModelTest::test_torchscript_output_attentions
> tests/test_modeling_bart.py::BartModelTest::test_torchscript_output_hidden_state
>   /home/nicolas/src/transformers/.venv/lib/python3.8/site-packages/torch/nn/functional.py:1897: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
>     assert padding_idx < weight.size(0), "Padding_idx must be within num_embeddings"
> 
> tests/test_modeling_bart.py::BartModelTest::test_torchscript
> tests/test_modeling_bart.py::BartModelTest::test_torchscript_output_attentions
> tests/test_modeling_bart.py::BartModelTest::test_torchscript_output_hidden_state
> tests/test_modeling_bart.py::BartStandaloneDecoderModelTest::test_torchscript
> tests/test_modeling_bart.py::BartStandaloneDecoderModelTest::test_torchscript_output_attentions
> tests/test_modeling_bart.py::BartStandaloneDecoderModelTest::test_torchscript_output_hidden_state
>   /home/nicolas/src/transformers/src/transformers/models/bart/modeling_bart.py:213: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
>     assert attn_weights.size() == (
> 
> tests/test_modeling_bart.py::BartModelTest::test_torchscript
> tests/test_modeling_bart.py::BartModelTest::test_torchscript_output_attentions
> tests/test_modeling_bart.py::BartModelTest::test_torchscript_output_hidden_state
> tests/test_modeling_bart.py::BartStandaloneDecoderModelTest::test_torchscript
> tests/test_modeling_bart.py::BartStandaloneDecoderModelTest::test_torchscript_output_attentions
> tests/test_modeling_bart.py::BartStandaloneDecoderModelTest::test_torchscript_output_hidden_state
>   /home/nicolas/src/transformers/src/transformers/models/bart/modeling_bart.py:220: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
>     assert attention_mask.size() == (
> 
> tests/test_modeling_bart.py::BartModelTest::test_torchscript
> tests/test_modeling_bart.py::BartModelTest::test_torchscript_output_attentions
> tests/test_modeling_bart.py::BartModelTest::test_torchscript_output_hidden_state
> tests/test_modeling_bart.py::BartStandaloneDecoderModelTest::test_torchscript
> tests/test_modeling_bart.py::BartStandaloneDecoderModelTest::test_torchscript_output_attentions
> tests/test_modeling_bart.py::BartStandaloneDecoderModelTest::test_torchscript_output_hidden_state
>   /home/nicolas/src/transformers/src/transformers/models/bart/modeling_bart.py:252: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
>     assert attn_output.size() == (
> 
> tests/test_modeling_bart.py::BartModelTest::test_torchscript
> tests/test_modeling_bart.py::BartModelTest::test_torchscript_output_attentions
> tests/test_modeling_bart.py::BartModelTest::test_torchscript_output_hidden_state
> tests/test_modeling_bart.py::BartStandaloneDecoderModelTest::test_torchscript
> tests/test_modeling_bart.py::BartStandaloneDecoderModelTest::test_torchscript_output_attentions
> tests/test_modeling_bart.py::BartStandaloneDecoderModelTest::test_torchscript_output_hidden_state
>   /home/nicolas/src/transformers/src/transformers/models/bart/modeling_bart.py:856: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
>     if input_shape[-1] > 1:
> 
> -- Docs: https://docs.pytest.org/en/stable/warnings.html
> ========================================= 106 passed, 8 skipped, 28 warnings in 562.15s (0:09:22) ==========================================
> ```

Ok for me then! If T5 tests pass this is good enough.

**Review by patil-suraj — APPROVED** [review]
Really cool design! LGTM

**Pull request merged**
